src/hotspot/os/linux/cgroupV1Subsystem_linux.cpp

```diff
@@ -88,6 +88,21 @@ void CgroupV1MemoryController::set_subsystem_path(char *cgroup_path) {
   }
 }
 
+/* read_mem_swap
+ *
+ * Returns the memory and swap metric. Note that calling code
+ * needs to handle re-casting for negative values.
+ *
+ * return:
+ *    A number > 0 if available,
+ *    OSCONTAINER_ERROR (as julong) for not supported
+ */
+julong CgroupV1Subsystem::read_mem_swap() {
+  GET_CONTAINER_INFO(julong, _memory->controller(), "/memory.memsw.limit_in_bytes",
+                     "Memory and Swap Limit is: ", JULONG_FORMAT, JULONG_FORMAT, memswlimit);
+  return memswlimit;
+}
```

Comment:
I think the API (and behavior) should following existing functions like:

```
jlong CgroupV1Subsystem::read_memory_limit_in_bytes()
```
Which returns:

- -2 when there was an error in determining the limit
-  -1 when there's no limit, or if the limit is larger than a reasonable physical maximum
- otherwise a non-negative number

That way, you won't need to do the awkward casting in the callers to check for OSCONTAINER_ERROR.

Since this is a recurring pattern, maybe this will be a good time to introduce a utility function for reading and capping the limit, as I suggested in https://github.com/openjdk/jdk/pull/12166#discussion_r1106387099 :-)

Reply:
> I think the API (and behavior) should following existing functions like:
> 
> ```
> jlong CgroupV1Subsystem::read_memory_limit_in_bytes()
> ```
> 
> Which returns:
> 
>     * -2 when there was an error in determining the limit
> 
>     * -1 when there's no limit, or if the limit is larger than a reasonable physical maximum
> 
>     * otherwise a non-negative number
> 
> 
> That way, you won't need to do the awkward casting in the callers to check for OSCONTAINER_ERROR.

Done in the update. We still need to distinguish between `OSCONTAINER_ERROR` returns and `-1` returns though. The former, resets it to the memory limit. The latter returns `-1`. At the same time, I've restructured how `swappiness` values are handled. IMO this makes it easier to read.

> Since this is a recurring pattern, maybe this will be a good time to introduce a utility function for reading and capping the limit, as I suggested in [#12166 (comment)](https://github.com/openjdk/jdk/pull/12166#discussion_r1106387099) :-)

Yes, we need to refactor the code. In particular, I'd like to get rid of the `GET_CONTAINER_INFO` macros as they're a pain to work with. I've filed https://bugs.openjdk.org/browse/JDK-8302744 to get this done.